worker/bootstrap/manifold.go

```diff
@@ -140,44 +189,38 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
 	}
 }
 
-// CAASAgentBinarySeeder is the function that is used to populate the tools
+// CAASAgentBinaryUploader is the function that is used to populate the tools
 // for CAAS.
-func CAASAgentBinarySeeder() error {
+func CAASAgentBinaryUploader(context.Context, string, BinaryAgentStorageService, objectstore.ObjectStore, Logger) error {
+	// CAAS doesn't need to populate the tools.
 	return nil
 }
 
-// IAASAgentBinarySeeder is the function that is used to populate the tools
+// IAASAgentBinaryUploader is the function that is used to populate the tools
 // for IAAS.
-func IAASAgentBinarySeeder() error {
-	return nil
-}
+func IAASAgentBinaryUploader(ctx context.Context, dataDir string, storageService BinaryAgentStorageService, objectStore objectstore.ObjectStore, logger Logger) error {
+	storage, err := storageService.AgentBinaryStorage(objectStore)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer storage.Close()
 
-// ApplicationService is the interface that is used to check if an application
-// exists.
-type ApplicationService interface {
-	ApplicationExists(name string) (bool, error)
+	return bootstrap.PopulateAgentBinary(ctx, dataDir, storage, logger)
 }
 
-// RequiresBootstrap returns true if the controller application does not exist.
-func RequiresBootstrap(appService ApplicationService) (bool, error) {
-	ok, err := appService.ApplicationExists(application.ControllerApplicationName)
-	if err != nil {
+// RequiresBootstrap returns true if the bootstrap params file exists.
+// It is expected at the end of bootstrap that the file is removed.
+func RequiresBootstrap(config agent.Config) (bool, error) {
```

Comment:
Wouldn't it be better to name this specifically, say `BootstrapParamsFileExists` with a comment as to what it is a proxy for?

Reply:
Yeah, originally we checked whether the application existed or not, so it made sense to be generalised about what we were waiting for. Now that we've focused in on the bootstrap file, I'm happy to change it.